### PR TITLE
Add workarounds for g++ 4.9

### PR DESCRIPTION
**Pull request by nguarracino**
This change uses some of the MSVC workarounds for older g++ compilers (gcc < 5.1) which have similar issues with constexpr.

Tests pass on:
- gcc-4.9.2/Linux
- gcc-5.2.1/Linux
- clang-3.6.1/Linux
- MSVC 2015/Windows

### CODE REVIEW

File: include/gsl.h [hunk 1]
```diff
@@ -45,6 +45,10 @@
 
 #endif // _MSC_VER
 
+#if (__GNUC__ < 5) && !defined(__clang__)
```
Comment (gdr-at-ms) [inline include/gsl.h]: You want to test first that `__GNUC__` is defined, via `defined(__GNUC__)`.

File: include/span.h [hunk 2]
```diff
@@ -60,6 +60,10 @@
 
 #endif // _MSC_VER
 
+#if (__GNUC__ < 5) && !defined(__clang__)
+#define constexpr /* nothing */
```
Comment (gdr-at-ms) [inline include/span.h]: Same comment here and elsewhere.

**Comment by gdr-at-ms** [comment]
We do really want to set the base line at C++14.

**Comment by nguarracino** [comment]
@gdr-at-ms I understand... problem is, I'm targeting an embedded device with a slightly older compiler. No worries if you want to keep the workarounds to a minimum, I can just keep this as a local patch and wait for my vendor to update to gcc 5.

**Comment by neilmacintosh** [comment]
@nguarracino I appreciate the effort (and interest in using the GSL!) but I'd prefer if you kept the PR in a fork (especially if you think it's likely you'll move to GCC 5 anyway). 

As per the comment in the Readme.md, our aim here is to rely on C++14-compliant compilers. We make an exception for MSVC for the reason that we have a lot of MSVC users internally . But even there we are applying strong pressure on its C++14 compliance so we can drop the workarounds as soon as possible.

**Pull request closed without merging**